src/library_vr.js

```diff
@@ -0,0 +1,227 @@
+var LibraryWebVR = {
+  $WebVR: {
+    DEVICE_TYPE_UNKNOWN: 0,
+    DEVICE_TYPE_HMD: 1,
+    DEVICE_TYPE_SENSOR: 2,
+
+    EYE_LEFT: 0,
+    EYE_RIGHT: 1,
+
+    initialized: false,
+    ready: false,
+    devices: [],
+    deviceHardwareIds: {},
+    nextHardwareDeviceId: 1,
+    getDevicesPromise: null,
+
+    selectedHMD: null,
+    selectedHMDId: 0,
+
+    init: function() {
+      if (WebVR.initialized)
+        return true;
+
+      WebVR.initialized = true;
+
+      if (!navigator.getVRDevices) {
+        WebVR.ready = true;
+        WebVR.devices = [];
+        return true;
+      }
+
+      navigator.getVRDevices().then(
+        function(devs) {
+          WebVR.ready = true;
+          WebVR.devices = devs;
+        },
+        function() {
+          WebVR.ready = true;
+          WebVR.devices = [];
+        }
+      );
+    },
+
+    getDeviceByID: function(deviceId) {
+      if (deviceId < 1 || deviceId > WebVR.devices.length) {
+        console.log("library_vr getDeviceByID error: " + (new Error()).stack);
+        throw "Invalid VR device ID given to Emscripten WebVR API function!";
+      }
+
+      return WebVR.devices[deviceId-1];
+    }
+  },
+
+  emscripten_vr_init: function() {
+    WebVR.init();
+  },
+
+  emscripten_vr_ready: function() {
+    return WebVR.ready ? 1 : 0;
+  },
+
+  emscripten_vr_count_devices: function() {
+    return WebVR.devices.length;
+  },
+
+  emscripten_vr_get_device_id: function(deviceIndex) {
+    if (deviceIndex < 0 || deviceIndex >= WebVR.devices.length)
+      return 0;
+
+    // we're doing to treat the device ID the same as the index + 1
+    return deviceIndex + 1;
+  },
+
+  emscripten_vr_get_device_hwid: function(deviceId) {
+    var dev = WebVR.getDeviceByID(deviceId);
+    if (!WebVR.deviceHardwareIds[dev.hardwareUnitId]) {
+      WebVR.deviceHardwareIds[dev.hardwareUnitId] = WebVR.nextHardwareDeviceId++;
+    }
+    return WebVR.deviceHardwareIds[dev.hardwareUnitId];
+  },
+
+  emscripten_vr_get_device_type: function(deviceId) {
+    var dev = WebVR.getDeviceByID(deviceId);
+    if (dev instanceof HMDVRDevice) {
+      return WebVR.DEVICE_TYPE_HMD;
+    }
+    if (dev instanceof PositionSensorVRDevice) {
+      return WebVR.DEVICE_TYPE_SENSOR;
+    }
+    return WebVR.DEVICE_TYPE_UNKNOWN;
+  },
+
+  emscripten_vr_select_hmd_device: function(deviceId) {
+    if (deviceId == 0) {
+      WebVR.selectedHMD = null;
+      WebVR.selectedHMDId = 0;
+      return 1;
+    }
+
+    var dev = WebVR.getDeviceByID(deviceId);
+    if (!(dev instanceof HMDVRDevice)) {
+      throw "Trying to call emscripten_vr_select_hmd_device on invalid or non-HMD device ID!";
+    }
+    WebVR.selectedHMD = dev;
+    WebVR.selectedHMDId = deviceId;
+    return 1;
+  },
+
+  emscripten_vr_get_selected_hmd_device: function() {
+    return WebVR.selectedHMDId;
+  },
+
+  emscripten_vr_hmd_get_eye_translation: function(deviceId, whichEye, translationPtr) {
+    var dev = WebVR.getDeviceByID(deviceId);
```

Comment:
A superficial note: in this kind of code pattern it would be good to check `translationPtr` for null before `makeSetValue`ing to that address (and perhaps return failure of type `EINVAL` or similar), since storing to null memory address silently works and does the wrong thing.